### ISSUE TITLE
Broken map reference on coordinator restart

### ISSUE THREAD

**Issue by gtfierro** [gtfierro/cs262-project#2]
If the coordinator has been running with some publisher (and likely a client too), and then crashes or is restarted, and then a client comes back _before_ the publisher, the coordinator crashes on [this line](https://github.com/gtfierro/cs262-project/blob/master/coordinator/forwarding_table.go#L385) with a nil map reference.

The issue, I believe, is that when the broker comes back online, the metadata has been persisted in its local MongoDB, so when the query is evaluated, it finds producers to forward. _However_, those producers have not yet re-registered with the broker (perhaps they have gone away), so the entries should be added to the forwarding table, but should be checked if they exist locally first.

**Comment by xkrogen** [comment]
Right... When I designed it I was kind of assuming the coordinator would never fail since I hadn't gotten to the fault tolerance portion of the coding yet. 

I think the right way to approach this may be to store all client/publisher information in the database (only the portions which can't be recovered automatically, i.e. client address / query mappings and publisher uuid / metadata mappings). This gives all of the coordinators all of the data they need to completely reconstruct the mapping table. But, we would prefer not to have to completely rebuild the mapping tables on coordinator failover, especially since that would beg the question of what SubscriptionDiff / ForwardRequest / etc. messages need to be sent during the coordinator recreation process... I'm beginning to think that maybe doing a RAFT round on each query/publication change may be best after all, and then each coordinator could simply maintain all of the tables locally and be immediately ready for failover. We are assuming that the rate of message transmission is much greater than the rate of metadata/query changes, after all. This would even allow us to do away with MongoDB replication; each could just  maintain their own local db (which would actually be replicated but wouldn't require the overhead beyond our own RAFT transactions). I dunno, kind of spitballing here, what do you think?

**Comment by gtfierro** [comment]
Well, the individual nodes still crash and come online even if the logical coordinator doesn't, so we do need to have them recover their state. I think you're right we should commit this to the database and have it pull it in on loading. 

I think we need some sort of "log replaying" mechanism. The master coordinator handles requests and makes the decisions, but it logs all of its actions (additions to, deletions from forwarding table -- I think everything except what is saved in the database) to some reliable log -- maybe stored in mongo -- that then gets replayed by other coordinators. They checkpoint their state and save it to mongo and then reload from the checkpoint and replay the log when they come back online. I think this is fairly similar to the kinds of systems we read about this semester.

How does that sound?

**Comment by gtfierro** [comment]
As an addition, this sounds pretty cool, so I don't mind implementing it if you don't have time.

**Comment by xkrogen** [comment]
Right, right, of course. Technically they should have a way to recover their state from another coordinator, not just their own local db, in case e.g. the machine actually failed rather than just transiently crashing. 

I think you're right - after I posted my last comment I was starting to think along the lines of needing to checkpoint and be able to replay. I'm not sure about storing a log or checkpointed state in Mongo - you're more familiar with it than I am, will it be amenable to that sort of thing? The checkpointed state would essentially be a big blog of serialized Go maps so as long as Mongo doesn't mind storing a big chunk of binary data (which, now that I think about it, should be fine, right?), and I guess I could see having the "log" just be a table where the primary key is basically the log sequence number... So maybe this does work. Then we just use RAFT for leader election, the leader is the only one who can write new log entries, etc. I should be able to implement it... This is pretty much the core of the distributed version of coordinators and I said I'd take that on so I'll try to tackle it tomorrow. I'll probably write up a bit more formalized description of what will happen first thing in the morning and send it over to you.

**Comment by gtfierro** [comment]
Sounds good!
